src/state/machines/turn.ts

```diff
@@ -281,28 +278,53 @@ export const TurnMachine = setup({
         draft.disasterMarket.deck = draft.disasterMarket.deck.slice(1);
       }),
     ),
+    stageHabitatUnlock: assign(({ context }: { context: GameState }) =>
+      produce(context, (draft) => {
+        const player = find(draft.players, { uid: draft.turn.player })!;
+
+        const playedAnimals =
+          (player.hand
+            .filter(({ uid }) => context.turn.playedCards.includes(uid))
+            .filter(({ type }) => type === "animal") as AnimalCard[]) ?? [];
+
+        const animalHabitatPairs = getAnimalHabitatPairs(playedAnimals);
+        const sharedHabitats = context.habitatMarket.deck.filter(
+          (marketHabitat) => !marketHabitat.isAcquired && getSharedHabitats(playedAnimals).includes(marketHabitat.name),
+        );
+
+        if (animalHabitatPairs.length === 0) return;
+
+        draft.stage = {
+          eventType: "habitatUnlock",
+          cause: flatten(animalHabitatPairs),
+          effect: sharedHabitats,
+        };
+
+        player.hand = without(player.hand, ...flatten(animalHabitatPairs));
```

Comment:
Hmm... so you are moving cards between player hand and stage?

Not sure it's the best approach. Maybe we could store uid's like we do for played/exhausted cards?

I understand that it might be a little bit easier on `positioner` side of things. I think I have an elegant solution for that. Let's have a chat later